tegra-bootloader-update.c

```diff
@@ -578,7 +578,22 @@ process_entry (bup_context_t *bupctx, int bootfd, int gptfd, struct update_entry
 		perror(ent->devname);
 		return -1;
 	}
-	erase_size = 512 * ((ent->length + 511) / 512);
+
+	erase_size = lseek(fd, 0, SEEK_END);
```

Comment:
I haven't done enough research into this myself yet but I'm wondering if this change could blow away uboot environment at the end of VER partitions on some old 32.x releases per https://forums.developer.nvidia.com/t/mmcblk0boot1-usage-at-address-4177408-and-u-boot-parameter-storage-space-availability/82113/13.  May just want to make sure this doesn't end up being used on old branches somehow.

Reply:
In my current setup the VER partition ended up using this part of the code:
```
	if (ent->part != NULL)
		return maybe_update_bootpart(bootfd, gptfd, ent, strcmp(ent->partname, "BCT") == 0, initialize, bctctx);
```
Within `maybe_update_bootpart` the entire partition gets erased as well. Do you mean that for older releases the VER partition was a full device rather than a partition?

Reply:
The code path being changed here is the one for non-QSPI/mmcblk0bootX partitions, so it shouldn't affect the VER vs. U-Boot env problem.

@nielsavonds  I'm curious about the conditions that triggered the bricking problem you're solving here.  Can you provide some more details?

Reply:
Sure! We did an update that decreased the size of our device tree with 12 bytes. When the Jetson Nano then rebooted, it logged "Failing over to KFS1", "Failing over to KFS2", and then it was dead. This was because the checksum of the DTB partition did not match the expected checksum.

What happens is that the erase_size for the new DTB is smaller than the actual size of the old DTB. Consequently, the last part of the old DTB remains written on the partition. The checksum obviously doesn't take these "leftover" bytes into account, so it ends up being incorrect.

I think you should be able to reproduce this as follows:
- Create a build and flash it
- Append at least 512 random non-zero bytes to the end of the device tree binary
- Create a bl_update_payload using that appended device tree
- Flash that bl_update_payload using tegra-bootloader-update. Everything will be fine since the size has increased.
- Create another bl_update_payload using the original device tree
- Flash that bl_update_payload using tegra-bootloader-update. The device will not boot anymore.